### PR TITLE
[Hook] Introduce some more strategies

### PR DESCRIPTION
**Pull request by nupplaphil**
FollowUp #13246

### CODE REVIEW

File: src/Core/Lock/Capabilities/ICanLock.php [hunk 1]
```diff
@@ -19,7 +19,7 @@
  *
  */
 
-namespace Friendica\Core\Lock\Capability;
+namespace Friendica\Core\Lock\Capabilities;
```
Comment (MrPetovan) [inline src/Core/Lock/Capabilities/ICanLock.php]: Why the change? What does the plural give you that the shorter singular didn't? I don't think anyone is assuming that there's only one capability in the Capability namespace.

Comment (nupplaphil) [reply src/Core/Lock/Capabilities/ICanLock.php]: hmm .. yes, you're right

I just wasn't sure if changing it from singular to plural or vice versa :D

Comment (MrPetovan) [reply src/Core/Lock/Capabilities/ICanLock.php]: Just go for the simpler term. 😄

File: src/Core/Cache/Factory/Cache.php [hunk 2]
```diff
@@ -92,7 +69,7 @@ public function __construct(BaseURL $baseURL, IManageConfigValues $config, Datab
 	 */
 	public function createDistributed(string $type = null): ICanCache
```
Comment (MrPetovan) [inline src/Core/Cache/Factory/Cache.php]: This is only used in one place and without the parameter, do we still need it? What about the APCU type check that's never hit?

File: src/Core/Cache/Type/APCuCache.php [hunk 3]
```diff
@@ -23,20 +23,19 @@
...
 use Friendica\Core\Cache\Capability\ICanCacheInMemory;
-use Friendica\Core\Cache\Enum\Type;
 use Friendica\Core\Cache\Exception\InvalidCacheDriverException;
 
 /**
  * APCu Cache.
  */
 class APCuCache extends AbstractCache implements ICanCacheInMemory
 {
+	public static $NAME = 'apcu';
```
Comment (MrPetovan) [inline src/Core/Cache/Type/APCuCache.php]: If it's not a constant, please use `snakeCase`.
```suggestion
	const NAME = 'apcu';
```
```suggestion
	public static $name = 'apcu';
```

File: src/Core/Cache/Factory/Cache.php [hunk 4]
```diff
@@ -92,7 +69,7 @@ public function __construct(BaseURL $baseURL, IManageConfigValues $config, Datab
...
 	{
-		if ($type === Enum\Type::APCU) {
+		if ($type === Type\APCuCache::$NAME) {
```
Comment (MrPetovan) [inline src/Core/Cache/Factory/Cache.php]: If the type check is warranted, why not check the class name instead?

File: static/strategies.config.php [hunk 5]
```diff
@@ -29,4 +31,18 @@
 		Type\SyslogLogger::class => ['syslog'],
 		Type\StreamLogger::class => ['stream'],
 	],
+	Cache\Capability\ICanCache::class => [
+		Cache\Type\APCuCache::class      => ['apcu'],
+		Cache\Type\DatabaseCache::class  => ['database', ''],
```
Comment (MrPetovan) [inline static/strategies.config.php]: Please define a constant with the value `'default'` that you can refer here and anywhere else you've used empty string to mean the default strategy.

The location and the name of the constant are left to you, although a mention of "default" would be nice.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: src/Core/Cache/Factory/Cache.php (hunk 4 above)
Comment (nupplaphil) [reply src/Core/Cache/Factory/Cache.php]: Because the class name isn't used as strategy name. It's the key of the mapping table of `strategies.config.php`

File: src/Core/Cache/Factory/Cache.php (hunk 2 above)
Comment (nupplaphil) [reply src/Core/Cache/Factory/Cache.php]: what do you mean? It's used here: https://github.com/friendica/friendica/blob/241a512589ac4deee398d922a2e14caea1639bbc/src/Core/Session/Factory/Session.php#L74

Comment (MrPetovan) [reply src/Core/Cache/Factory/Cache.php]: Yes, without parameter, does it still make sense to have one?

Comment (nupplaphil) [reply src/Core/Cache/Factory/Cache.php]: Yes you're right.. I now understand :D

File: src/Core/KeyValueStorage/Factory/KeyValueStorage.php [hunk 6]
```diff
@@ -19,17 +19,16 @@
...
-	const APCU      = 'apcu';
-	const REDIS     = 'redis';
-	const ARRAY     = 'array';
-	const MEMCACHE  = 'memcache';
-	const DATABASE  = 'database';
-	const MEMCACHED = 'memcached';
+	public function create(ICanCreateInstances $instanceCreator): IManageKeyValuePairs
+	{
+		/** @var IManageKeyValuePairs */
+		return $instanceCreator->create(IManageKeyValuePairs::class, '');
```
Comment (MrPetovan) [inline src/Core/KeyValueStorage/Factory/KeyValueStorage.php]: I believe the STRATEGY_DEFAULT_KEY constant should be used here.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Review by MrPetovan — APPROVED** [review]
